text/3851-supertrait-auto-impls.md

```diff
@@ -0,0 +1,975 @@
+trait Subtrait: Supertrait {
+    // Refactored into the Supertrait
+    //*** fn supertrait_fn();
+    fn subtrait_fn();
+}
+
+// The proposal called for enabling specialisation of
+// the following blanket implementation as *stop-gap*.
+// We propose that for trait evolution, we do not need to rely on specialisation...
+/**
+impl<T: Subtrait> Supertrait for T {
+    default fn supertrait_fn() {
+        <T as Subtrait>::supertrait_fn();
+    }
+}
+**/
+
+struct Middeware<T>(T);
+
+// ... but rather keep the current and future `impl` block
+// the same,
+impl Subtrait for Middleware<T> {
+    // because the trait has been well designed so that
+    // it allows seamless migration.
+    fn supertrait_fn() { .. }
+    fn subtrait_fn() { .. }
+}
+```
+
+
+
+## Default auto implementations
+
+It is possible to provide default implementations of functions from the super trait.
+```rs
+trait MyTrait {
+    fn my_func(&self);
+}
+
+trait MySubTrait: MyTrait {
+    auto impl MyTrait {
+        fn my_func(&self) {
+            self.my_second_func();
+            self.my_second_func();
+        }
+    }
+    
+    fn my_second_func(&self);
+}
+```
+In this case, an impl block for `MySubTrait` will still automatically implement `MyTrait`, but you are not required to provide an implementation of `my_func` since a default implementation exists.
+
+### Example: Helpers for implementing a trait
+
+With default auto implementations, it becomes possible to provide a sub-trait whose purpose is to help you implement the super trait in a specific way.
+
+For example, given this super trait:
+```rs
+trait EventHandler {
+    type Event;
+    fn handle_event(&mut self, event: Self::Event);
+}
+```
+Then you might have a helper for implementing `EventHandler` for a specific event type:
+```rs
+enum MouseEvent {
+    ClickEvent(ClickEvent),
+    MoveEvent(MoveEvent),
+}
+
+trait MouseEventHandler: EventHandler {
+    auto impl EventHandler {
+        type Event = MouseEvent;
+        fn handle_event(&mut self, event: MouseEvent) {
+            use MouseEvent::*;
+            match event {
+                ClickEvent(evt) => self.click_event(evt),
+                MoveEvent(evt) => self.move_event(evt),
+            }
+        }
+    }
+    
+    fn click_event(&mut self, event: ClickEvent);
+    fn move_event(&mut self, event: MoveEvent);
+}
+```
+This allows crates to implement `EventHandler` by writing this code:
+```rs
+struct PrintHandler;
+
+impl MouseEventHandler for PrintHandler {
+    fn click_event(&mut self, event: ClickEvent) {
+        println!("Click: {:?}", event);
+    }
+    fn move_event(&mut self, event: MoveEvent) {
+        println!("Move: {:?}", event);
+    }
+}
+```
+The `MouseEventHandler` trait could even come from a different crate than `EventHandler`.
+
+## Unsafe auto impl
+
+It's possible to declare that an auto implementation is unsafe.
+```rs
+trait MySubTrait: MyTrait {
+    unsafe auto impl MyTrait;
```

Comment:
so, does this create an `unsafe` requirement or discharge an existing `unsafe` requirement? you seem to have it do both which is generally not how Rust uses `unsafe` anymore, now that we have [`unsafe_op_in_unsafe_fn`](https://doc.rust-lang.org/edition-guide/rust-2024/unsafe-op-in-unsafe-fn.html) enabled by default

Reply:
I see that this can be confusing.

- Inside a `trait Subtrait: .. { .. }` block, `unsafe auto impl` creates an `unsafe` requirement.
- Inside a `impl Subtrait { .. }` block, `unsafe auto impl` discharges the corresponding `unsafe` requirement.

Should we adjust the "keyword" orders, so that the difference can be more obvious?

Reply:
+1 to @programmerjake 's expectations-- I would've expected `unsafe auto impl` to discharge an unsafe requirement-- that is, guarante that the provided impl of the supertrait fulfills the requirements of the `unsafe trait SuperTrait`. This is the usual meaning of `unsafe impl` in Rust.